messages/messages.go

```diff
@@ -0,0 +1,126 @@
+)
+
+const (
+	// sha1Prefix is the prefix used by Github before the HMAC hexdigest.
+	sha1Prefix = "sha1="
+	// sha256Prefix and sha512Prefix are provided for future compatibility.
+	sha256Prefix = "sha256="
+	sha512Prefix = "sha512="
+	// eventHeader is the Github header key used to pass the event type.
+	eventHeader = "X-Github-Event"
+	// signatureHeader is the Github header key used to pass the HMAC hexdigest.
+	signatureHeader = "X-Hub-Signature"
+	// deliveryHeader is the Gihub header key used to pass the globally-unique Github Event ID.
+	deliveryHeader = "X-GitHub-Delivery"
+)
+
+// GenSignature generates a signature for the given payload.
+// hashType possible values are "sha1", "sha256", "sha512".
+// If hashType is empty or invalid, "sha1" is used.
+func GenSignature(payload, secretKey []byte, hashType string) string {
+	var hashFunc func() hash.Hash
+	prefix := sha1Prefix
+	switch hashType {
+	case "sha256":
+		hashFunc = sha256.New
+		prefix = sha256Prefix
+	case "sha512":
+		hashFunc = sha512.New
+		prefix = sha512Prefix
+	}
+	return prefix + hex.EncodeToString(genMAC(payload, secretKey, hashFunc))
+}
+
+// genMAC generates the HMAC signature for a message provided the secret key and hashFunc.
+// If hashFunc is nil, sha1.New is used.
+func genMAC(message, key []byte, hashFunc func() hash.Hash) []byte {
+	if hashFunc == nil {
+		hashFunc = sha1.New
+	}
+	mac := hmac.New(hashFunc, key)
+	mac.Write(message)
+	return mac.Sum(nil)
+}
+
+// checkMAC reports whether messageMAC is a valid HMAC tag for message.
+// If hashFunc is nil, sha1.New is used.
+func checkMAC(message, messageMAC, key []byte, hashFunc func() hash.Hash) bool {
+	expectedMAC := genMAC(message, key, hashFunc)
+	return hmac.Equal(messageMAC, expectedMAC)
+}
+
+// EventPayload validates an incoming Github Webhook event request
+// and returns the event ID, event type, and the (JSON) payload.
+// secretKey is the GitHub Webhook secret message.
+func EventPayload(r *http.Request, secretKey []byte) (eventID, event string, payload []byte, err error) {
+	payload, err = ioutil.ReadAll(r.Body)
+	if err != nil {
+		return "", "", nil, err
+	}
+
+	sig := r.Header.Get(signatureHeader)
+	if err := ValidateSignature(sig, payload, secretKey); err != nil {
+		return "", "", nil, err
+	}
+
+	eventID = r.Header.Get(deliveryHeader)
+	if event = r.Header.Get(eventHeader); event == "" {
+		return "", "", nil, errors.New("missing event type")
+	}
+	return eventID, event, payload, nil
+}
+
+// ValidateSignature validates the signature for the given payload.
+// sig is the GitHub hash signature delivered in the X-Hub-Signature header.
+// payload is the JSON payload sent by GitHub Webhooks.
+// secretKey is the GitHub Webhook secret message.
+// GitHub docs: https://developer.github.com/webhooks/securing/#validating-payloads-from-github
+func ValidateSignature(sig string, payload, secretKey []byte) error {
+	if sig == "" {
+		return errors.New("missing signature")
+	}
+	var sigLen int
```

Comment:
sigLen is a slightly confusing name, since this isn't actually the length of the signature, it's the length of the prefix.  An alternate option (which might be easier to read, I'm not sure?) would be to:

```go
sigParts := strings.SplitN(sig, "=", 2)
...
```